src/backend/distributed/connection/connection_management.c

```diff
@@ -438,183 +479,377 @@ ShutdownConnection(MultiConnection *connection)
+				ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
 			}
 
-			pollResult = (select) (selectFileDescriptor + 1, &readFileDescriptorSet,
-								   &writeFileDescriptorSet, &exceptionFileDescriptorSet,
-								   &selectTimeout);
-#endif /* HAVE_POLL */
-
-			if (pollResult == 0)
+			if (event->events & WL_LATCH_SET)
 			{
-				/*
-				 * Timeout exceeded. Two things to do:
-				 * - check whether any interrupts arrived and handle them
-				 * - check whether establishment for connection already has
-				 *   lasted for too long, stop waiting if so.
-				 */
+				ResetLatch(MyLatch);
+
 				CHECK_FOR_INTERRUPTS();
 
-				if (TimestampDifferenceExceeds(connection->connectionStart,
-											   GetCurrentTimestamp(),
-											   NodeConnectionTimeout))
+				if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
 				{
-					ereport(WARNING, (errmsg("could not establish connection after %u ms",
-											 NodeConnectionTimeout)));
-
-					/* close connection, otherwise we take up resource on the other side */
-					PQfinish(connection->pgConn);
-					connection->pgConn = NULL;
-					break;
+					/*
+					 * because we can't break from 2 loops easily we need to not forget to
+					 * reset the memory context
+					 */
+					MemoryContextDelete(MemoryContextSwitchTo(oldContext));
+					return;
 				}
+
+				continue;
 			}
-			else if (pollResult > 0)
-			{
-				/*
-				 * IO possible, continue connection establishment. We could
-				 * check for timeouts here as well, but if there's progress
-				 * there seems little point.
-				 */
-				break;
-			}
-			else
+
+			switch (MultiConnectionStatePoll(connectionState))
 			{
-				int errorCode = errno;
-#ifdef WIN32
-				errorCode = WSAGetLastError();
-#endif
-				if (errorCode == EINTR)
+				case REBUILD:
 				{
-					/* Retrying, signal interrupted. So check. */
-					CHECK_FOR_INTERRUPTS();
+					waitEventSetRebuild = true;
+					break;
 				}
-				else
+
+				case UPDATE:
 				{
-					/*
-					 * We ERROR here, instead of just returning a failed
-					 * connection, because this shouldn't happen, and indicates a
-					 * programming error somewhere, not a network etc. issue.
-					 */
-					ereport(ERROR, (errcode_for_socket_access(),
-									errmsg("poll()/select() failed: %m")));
+					uint32 eventMask = MultiConnectionStateEventMask(connectionState);
+					ModifyWaitEvent(waitEventSet, event->pos, eventMask, NULL);
+					break;
 				}
+
+				case NONE:
+				{
+					break;
+				}
+			}
+		}
+
+		if (eventCount == 0)
+		{
+			/*
+			 * timeout has occured on waitset, double check the timeout since
+			 * connectionStart and if passed close all non-finished connections
+			 */
+
+			TimestampTz now = GetCurrentTimestamp();
+			if (TimestampDifferenceExceeds(connectionStart, now, NodeConnectionTimeout))
+			{
+				CloseNotReadyMultiConnectionStates(connectionStates);
+				ereport(ERROR, (errmsg("could not establish connection after %u ms",
```

Comment:
1. Previously we were emitting this message as a `WANRING`, why did we change it to an `ERROR`?
2. If this is intended to be an `ERROR`, do we have need to reset the temporary memory context before emitting the error?

Reply:
An explanation is on the previous discussion: https://github.com/citusdata/citus/pull/2584#discussion_r266419735
TLDR: not reporting it as an error causes a segmentation fault.

Connections are already added to the cache so we nee to call `CloseConnection` on all connections to remove them from the cache. `CloseConnection` free's the data of the `MultiConnection`. Functions calling `FinishConnection(List)Establishment` assume the connections are established.

By changing it to an error we break out from that codepath which makes it safe to call `CloseConnection`, not reporting it as an error causes a segmentation fault.

Besides removing the connection from the cache, the error message becomes more clear as the timeout is not reported as an error opposed to a generic `connection error` from the coordinator with a warning above hinting that a timeout occurred.

Reply:
Ok, makes sense. Thanks for the explanation. What's your take on my 2nd comment? i.e. calling before reporting errors `MemoryContextDelete(MemoryContextSwitchTo(oldContext));`? Will throwing an error automatically take care of that?

Reply:
Yes, assuming this function is called within `TopTransactionContext` the memory context will be deleted on abort.

Reply:
I was trying to look for documentation around this, but was hard to find. Thanks Marco for the explanation.

I did have a look at the implementation of `ereport(ERROR,..)` and found that one of the first things they do is switching memory contexts to the ErrorContext, so I guess we don't even need to switch the context back to the `oldContext` before calling.